pluto/worker.go

```diff
@@ -0,0 +1,99 @@
+package pluto
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync/atomic"
+)
+
+type worker struct {
+	begin   uint64
+	end     uint64
+	url     *url.URL
+	writer  io.WriterAt
+	headers []string
+	verbose bool
+	ctx     context.Context
+}
+
+// copyAt reads 64 kilobytes from source and copies them to destination at a given offset
+func (w *worker) copyAt(src io.Reader, dlcounter *uint64) (uint64, error) {
+	bufBytes := make([]byte, 256*1024)
+
+	var bytesWritten uint64
+	var err error
+
+	for {
+		nsr, serr := src.Read(bufBytes)
+		if nsr > 0 {
+			ndw, derr := w.writer.WriteAt(bufBytes[:nsr], int64(w.begin))
+			if ndw > 0 {
+				u64ndw := uint64(ndw)
+				w.begin += u64ndw
+				bytesWritten += u64ndw
+				atomic.AddUint64(dlcounter, u64ndw)
+			}
+			if derr != nil {
+				err = derr
+				break
+			}
+			if nsr != ndw {
+				fmt.Printf("Short write error. Read: %d, Wrote: %d", nsr, ndw)
+				err = io.ErrShortWrite
+				break
+			}
+		}
+
+		if serr != nil {
+			if serr != io.EOF {
+				err = serr
+			}
+			break
+		}
+	}
+
+	return bytesWritten, err
+}
+
+func (w *worker) download() (io.ReadCloser, error) {
+
+	client := &http.Client{}
+	req, err := http.NewRequest("GET", w.url.String(), nil)
+	if err != nil {
+		return nil, fmt.Errorf("error in creating GET request: %v", err)
+	}
+	req = req.WithContext(w.ctx)
+	req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", w.begin, w.end))
+
+	for _, v := range w.headers {
+		vsp := strings.Index(v, ":")
+
+		key := v[:vsp]
+		value := v[vsp:]
+
+		req.Header.Set(key, value)
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+
+		if w.verbose {
+			fmt.Printf("Requested Bytes %d in range %d-%d. Got %d bytes\n", w.end-w.begin, w.begin, w.end, resp.ContentLength)
+		}
+
+		return nil, fmt.Errorf("error in sending download request: %v", err)
+	}
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusPartialContent {
+		return nil, fmt.Errorf("status code: %d", resp.StatusCode)
+	}
+
+	/*if uint64(resp.ContentLength) != (w.end - w.begin) {
```

Comment:
This check is necessary as several caches like Varnish don't obey the RFC correctly and send the complete file instead of the requested data. Without this check It'll corrupt the file that is being downloaded. 
To test this, Run a simple python http server and try downloading a file from that server using pluto.

Reply:
I was wondering why there was this check. In the tests I did I found that some servers answer sometimes 1byte more. I wanted to check a little more why this happens (this is why I commented out the code and not removed it). I will let is as it was

Reply:
@josledp 1 or 2 versions of Varnish have this bug. 

In a scenario when you are downloading a file of say 10 bytes with 3 parts. Most servers send the last remaining byte in the request which requested bytes from 6-9. This is slightly annoying because there are more servers which send 2-3 bytes of remaining data with the last request. 

This is probably done to prevent the HTTP overhead in sending 2-3 bytes of data but I am not sure where to draw the line? When should it start treating extra data sent by server as corrupted data? 

BTW, Are you still working on this? Apolgoies for not replying sooner.